### PR TITLE
tools: Refactor CI to use `argh`

### PR DESCRIPTION
**Pull request by MrGVSV**
# Objective

The CI tool currently parses input manually. This has worked fine, but makes it just a bit more difficult to maintain and extend. Additionally, it provides no usage help for devs wanting to run the tool locally.

It would be better if parsing was handled by a dedicated CLI library like [`clap`](https://github.com/clap-rs/clap) or [`argh`](https://github.com/google/argh).

## Solution

Use `argh` to parse command line input for CI.

`argh` was chosen over `clap` and other tools due to being more lightweight and already existing in our dependency tree.

Using `argh`, the usage notes are generated automatically:

```
$ cargo run -p ci --quiet -- --help
Usage: ci [--keep-going] [<command>] [<args>]

The CI command line tool for Bevy.

Options:
  --keep-going      continue running commands even if one fails
  --help            display usage information

Commands:
  lints             Alias for running the `format` and `clippy` subcommands.
  doc               Alias for running the `doc-test` and `doc-check`
                    subcommands.
  compile           Alias for running the `compile-fail`, `bench-check`,
                    `example-check`, `compile-check`, and `test-check`
                    subcommands.
  format            Check code formatting.
  clippy            Check for clippy warnings and errors.
  test              Runs all tests (except for doc tests).
  test-check        Checks that all tests compile.
  doc-check         Checks that all docs compile.
  doc-test          Runs all doc tests.
  compile-check     Checks that the project compiles.
  cfg-check         Checks that the project compiles using the nightly compiler
                    with cfg checks enabled.
  compile-fail      Runs the compile-fail tests.
  bench-check       Checks that the benches compile.
  example-check     Checks that the examples compile.
```

This PR makes each subcommand more modular, allowing them to be called from other subcommands. This also makes it much easier to extract them out of `main.rs` and into their own dedicated modules.

Additionally, this PR improves failure output:

```
$ cargo run -p ci -- lints
...
One or more CI commands failed:
format: Please run 'cargo fmt --all' to format your code.
```

Including when run with the `--keep-going` flag:

```
$ cargo run -p ci -- --keep-going lints
...
One or more CI commands failed:
- format: Please run 'cargo fmt --all' to format your code.
- clippy: Please fix clippy errors in output above.
```

### Future Work

There are a lot of other things we could possibly clean up. I chose to try and keep the API surface as unchanged as I could (for this PR at least).

For example, now that each subcommand is an actual command, we can specify custom arguments for each.

The `format` subcommand could include a `--check` (making the default fun `cargo fmt` as normal). Or the `compile-fail` subcommand could include `--ecs`, `--reflect`, and `--macros` flags for specifying which set of compile fail tests to run.

The `--keep-going` flag could be split so that it doesn't do double-duty where it also enables `--no-fail-fast` for certain commands. Or at least make it more explicit via renaming or using alternative flags.

---

## Changelog

- Improved the CI CLI tool
  - Now includes usage info with the `--help` option!
- [Internal] Cleaned up and refactored the `tools/ci` crate using the `argh` crate

## Migration Guide

The CI tool no longer supports running multiple subcommands in a single call. Users who are currently doing so will need to split them across multiple commands:

```bash
# BEFORE
cargo run -p ci -- lints doc compile

# AFTER
cargo run -p ci -- lints && cargo run -p ci -- doc && cargo run -p ci -- compile
# or
cargo run -p ci -- lints; cargo run -p ci -- doc; cargo run -p ci -- compile
# or
cargo run -p ci -- lints
cargo run -p ci -- doc
cargo run -p ci -- compile
```

### CODE REVIEW

**Review by BD103 — CHANGES_REQUESTED** [review]
I love this change. My one complaint is that I don't like how `keep_going` is duplicated across every single subcommand and the main `CI` struct. I see a few alternatives:

1. Only accept `--keep-going` before subcommands, removing it as a field from the subcommand structs. I prefer this, since all commands appear to support `--no-fail-fast`.*
2. Only accept `--keep-going` after subcommands, removing it from the global `CI` logic and only keeping it for commands that actually implement the feature. This would mean a lot of duplicated code, but it would allow us to not support `--keep-going` for commands that do not support it.

*The Clippy check is the only subcommand in the code that does not check for `--keep-going`. I think this is a mistake, since Clippy inherits all options from `cargo check`, which **does** support `--no-fail-fast`.

**Comment by MrGVSV** [comment]
> I love this change. My one complaint is that I don't like how `keep_going` is duplicated across every single subcommand and the main `CI` struct.

Agreed! I only kept it like this to minimize the breakage introduced in this particular PR. Ideally, we wouldn't have to do this.

> 1. Only accept `--keep-going` before subcommands, removing it as a field from the subcommand structs. I prefer this, since all commands appear to support `--no-fail-fast`.*

I think I prefer this as well. It allows each command to specify its own options like `--no-fail-fast`, which are both more explicit and localized to the actual command being run. 

And from a quick search in the code base, I couldn't see anywhere where `--keep-going` is actually being used, so it's probably fine to do something like this.

I think such a change might be better for a followup PR so that it gets more focus and doesn't block this change. Thoughts on that? Or should I include the change now?

**Comment by BD103** [comment]
> And from a quick search in the code base, I couldn't see anywhere where `--keep-going` is actually being used, so it's probably fine to do something like this.
> 
> I think such a change might be better for a followup PR so that it gets more focus and doesn't block this change. Thoughts on that? Or should I include the change now?

I think it is fine to do this in the same PR, since your already changing it so multiple subcommands need to be run separately. (Unfortunately I'm a bad example, since I end up just skipping `ci` and running Clippy and rustfmt directly. Maybe ask a maintainer what their workflow is?)

I believe there was some talk about using the `--keep-going` flag in Github Actions, but I don't know if it was actually added. Ripgrep doesn't show anything.

**Comment by MrGVSV** [comment]
> I think it is fine to do this in the same PR, since your already changing it so multiple subcommands need to be run separately.

Went ahead and removed the hidden `--keep-going` flags!

> The Clippy check is the only subcommand in the code that does not check for `--keep-going`. I think this is a mistake, since Clippy inherits all options from `cargo check`, which **does** support `--no-fail-fast`.

I was looking into this as well and from what I can tell (running `cargo check --help` anyways), it doesn't seem like `cargo check` supports `--no-fail-fast`. It actually looks like only `cargo test` might support this flag. I wonder if we should also remove the places that use it unnecessarily?

File: tools/ci/src/commands/subcommands/bench_check_command.rs
Comment (BD103) [inline tools/ci/src/commands/subcommands/bench_check_command.rs:8]:
```suggestion
pub(crate) struct BenchCheckCommand;
```

If the subcommands don't contain any fields, they can be converted to unit structs.

Beyond that, this looks good to go! (Don't worry about CI, btw, it's failing because of our dependency on Taffy.)

Comment (MrGVSV) [reply tools/ci/src/commands/subcommands/bench_check_command.rs:8]: I believe `argh` disallows unit structs. I can double check, though

Comment (BD103) [reply tools/ci/src/commands/subcommands/bench_check_command.rs:8]: I tested it, your correct! Good to know for the future

**Review by BD103 — APPROVED** [review]
Thanks! I already have 3 different ideas for how to improve the `ci` tool using these changes :)

**Review by alice-i-cecile — APPROVED** [review]
I'm a big fan of the nicer argument parsing, and of the new file structure. Good comments too.

**Pull request merged**
